### ISSUE TITLE
Segmentation fault on invalid timeout during eval inside a try/catch

### ISSUE THREAD

**Issue by josevalim** [erlang/otp#5235]
**Describe the bug**

Easiest way to explain this one is with code. Write this module:

```erlang
-module(foo).
-export([run/0, eval/0]).

run() ->
  try
    eval()
  catch
    _:_:Stacktrace ->
      io:format("~p~n", [Stacktrace])
  end.

eval() ->
  Send = {call,1,
      {remote,1,{atom,1,erlang},{atom,1,send}},
      [{call,1,{remote,1,{atom,1,erlang},{atom,1,self}},[]},{atom,1,hello}]},
  erl_eval:expr(Send, #{}, none, none, none),

  Receive = {'receive',3,
           [{clause,4,[{atom,4,none}],[],[{atom,4,never}]}],
           {atom,6,nil},
           [{atom,6,never}]},
  erl_eval:expr(Receive, #{}, none, none, none).
```

Now run it:

```
$ erlc foo.erl && erl -s foo run
Erlang/OTP 24 [erts-12.0] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit]

Segmentation fault: 11
```

**Affected versions**

I can reproduce this on OTP 24 and OTP 24.0.6. I am building 24.1 right now to see if it still exists there.

EDIT: the issue exists on 24.1.

**Comment by rickard-green** [comment]
Thanks for the bug report! I've made a PR #5238 (based on OTP 24.0) that should fix this issue. Note that it has only passed basic testing yet.

**Comment by rickard-green** [comment]
The crash was just the effect of another bug which left the message queue in an inconsistent state. I suspect this bug has been present since the birth of Erlang.

The process have a pointer which keeps track of the next message to inspect. When leaving a receive, this pointer should be reset to the beginning of the message queue. When leaving a receive due to a faulty timeout value, this was not done. If this exception is caught, the next receive wont search from the beginning of the message queue, but instead from the last message searched in the previous receive. This can for example cause the process to hang forever in a receive even if we have a matching message in the message queue:

```
    self() ! init,
    try
        receive blipp -> ok after blupp -> ok end
    catch _:_ ->
            ok
    end,
    receive init -> ok end.
```

I've made a new PR #5241 with a fix of this bug. It also contains the previous fix, which most likely is unnecessary. Note that also this fix only has gone through very basic testing.

**Comment by rickard-green** [comment]
PR replaced by yet another PR #5244 due to an unfortunate branch rename :-/

**Comment by rickard-green** [comment]
Now fixed in the just released [OTP 24.1.1](https://erlang.org/download/OTP-24.1.1.README) patch. The fix will be released in the next patches on OTP 22 and OTP 23 as well.
